src/lib/salad/hll.c

```diff
@@ -0,0 +1,954 @@
+ * merging, o(buffer_size + list_size) to merge and o(buffer_size + list_size)
+ * to remove duplicate indexes.
+ */
+static void
+hll_sparse_merge_list_with_buffer(struct hll *hll)
+{
+	if (hll_sparse_pairs_buffer_size(hll->data) == 0)
+		return;
+
+	hll_sparse_sort_buff(hll);
+
+	pair_t *list = hll_sparse_pairs_list_begin(hll->data);
+	pair_t *list_end = hll_sparse_pairs_list_end(hll->data);
+	pair_t *buff = hll_sparse_pairs_buffer_begin(hll->data);
+	pair_t *buff_end = hll_sparse_pairs_buffer_end(hll->data);
+	assert(list_end <= buff);
+
+	uint8_t *new_pairs = xcalloc(hll_sparse_bsize(hll->data), 1);
+	pair_t *new_list = hll_sparse_pairs_list_begin(new_pairs);
+	uint32_t new_list_size = merge_sorted_pairs_arrays(list, list_end,
+							   buff, buff_end,
+							   new_list);
+	new_list_size =
+		merge_pairs_with_duplicate_indexes(new_list, new_list_size);
+
+	struct pairs_header *new_header = hll_sparse_pairs_header(new_pairs);
+	new_header->bsize = hll_sparse_bsize(hll->data);
+	new_header->list_idx = new_list_size;
+	new_header->buff_idx = hll_sparse_pairs_max_size(hll->data);
+	free(hll->data);
+	hll->data = new_pairs;
+}
+
+/* Add a new pair to the densely represented HyperLogLog. */
+static void
+hll_dense_add_pair(struct hll *hll, pair_t pair)
+{
+	uint32_t idx = hll_sparse_pair_dense_idx(pair, hll->dprecision);
+	uint8_t new_rank = hll_sparse_pair_dense_rank(pair);
+	hll_dense_add_to_register(hll, idx, new_rank);
+}
+
+/* Add an array of pairs to the densely represented HyperLogLog. */
+static void
+hll_dense_add_pairs(struct hll *hll, pair_t *beg, pair_t *end)
+{
+	for (pair_t *pair = beg; pair != end; ++pair)
+		hll_dense_add_pair(hll, *pair);
+}
+
+/*
+ * Convert a sparsely represented HyperLogLog to a densely represented
+ * HyperLogLog. The sparsely represented HyperLogLog is freed after converting.
+ */
+static void
+hll_sparse_to_dense(struct hll *sparse_hll)
+{
+	assert(sparse_hll->representation == HLL_SPARSE);
+	hll_sparse_merge_list_with_buffer(sparse_hll);
+
+	uint8_t precision = sparse_hll->dprecision;
+	struct hll tmp;
+	struct hll *dense_hll = &tmp;
+	hll_create(dense_hll, precision, HLL_DENSE);
+
+	pair_t *list_beg = hll_sparse_pairs_list_begin(sparse_hll->data);
+	pair_t *list_end = hll_sparse_pairs_list_end(sparse_hll->data);
+	hll_dense_add_pairs(dense_hll, list_beg, list_end);
+	pair_t *buff_beg = hll_sparse_pairs_buffer_begin(sparse_hll->data);
+	pair_t *buff_end = hll_sparse_pairs_buffer_end(sparse_hll->data);
+	hll_dense_add_pairs(dense_hll, buff_beg, buff_end);
+
+	hll_destroy(sparse_hll);
+	memcpy(sparse_hll, dense_hll, sizeof(*sparse_hll));
+}
+
+/*
+ * Add hash to the sparsely represented HyperLogLog estimator.
+ * The representation may be changed to dense after the call.
+ */
+static void
+hll_sparse_add(struct hll *hll, uint64_t hash)
+{
+	if (hll_sparse_is_full(hll))
+		hll_sparse_merge_list_with_buffer(hll);
+
+	if (hll_sparse_is_full(hll)) {
```

Comment:
Please move `hll_sparse_merge_list_with_buffer(hll);` here.

Reply:
Why? After the merge, the hll may not be filled.